plugins/outputs/mqtt/topic_name_generator.go

```diff
@@ -31,17 +36,39 @@ func NewTopicNameGenerator(topicPrefix, topic string) (*TopicNameGenerator, erro
 	return &TopicNameGenerator{TopicPrefix: topicPrefix, template: tt}, nil
 }
 
+func (t *TopicNameGenerator) Name() string {
+	return t.metric.Name()
+}
+
 func (t *TopicNameGenerator) Tag(key string) string {
-	tagString, _ := t.metric.GetTag(key)
-	return tagString
+	return t.metric.Tag(key)
+}
+
+func (t *TopicNameGenerator) Field(key string) interface{} {
+	return t.metric.Field(key)
+}
+
+func (t *TopicNameGenerator) Time() time.Time {
+	return t.metric.Time()
 }
 
-func (t *TopicNameGenerator) Generate(hostname string, m telegraf.Metric) (string, error) {
-	t.Hostname = hostname
-	t.metric = m
-	t.PluginName = m.Name()
+func (t *TopicNameGenerator) Tags() map[string]string {
+	return t.metric.Tags()
+}
+
+func (t *TopicNameGenerator) Fields() map[string]interface{} {
+	return t.metric.Fields()
+}
+
+func (t *TopicNameGenerator) String() string {
+	return t.metric.String()
+}
+
+func (m *MQTT) generateTopic(metric telegraf.Metric) (string, error) {
+	m.generator.metric = metric.(telegraf.TemplateMetric)
+
 	var b strings.Builder
-	err := t.template.Execute(&b, t)
+	err := m.generator.template.Execute(&b, m.generator)
```

Comment:
Wouldn't passing in `m.generator.metric.(telegraf.TemplateMetric)` work? If so we could get rid of all the `TopicNameGeneration` wrapper functions...

Reply:
No, it still uses `m.generator.TopicPrefix` as well. So as said in description this can be removed when `TopicPrefix` will be actually removed in v1.35.0. These changes ease that transition.